dds/DCPS/transport/rtps_udp/RtpsUdpDataLink.cpp

```diff
@@ -2803,16 +2801,14 @@ RtpsUdpDataLink::bundle_and_send_submessages(MetaSubmessageVec& meta_submessages
               OPENDDS_ASSERT(mapping.next_directed_unassigned_ != mapping.map_.end());
               map_pair.new_ = mapping.next_directed_unassigned_->first;
               ++mapping.next_directed_unassigned_;
-              if (res.sm_.heartbeat_sm().smHeader.flags & RTPS::OPENDDS_FLAG_R) {
-                if (res.sm_.heartbeat_sm().count.value != map_pair.new_) {
-                  update_required_acknack_count(res.src_guid_, res.dst_guid_, res.sm_.heartbeat_sm().count.value, map_pair.new_);
-                }
-                res.sm_.heartbeat_sm().smHeader.flags &= ~RTPS::OPENDDS_FLAG_R;
-              }
             }
             map_pair.is_new_assigned_ = true;
           }
           res.sm_.heartbeat_sm().count.value = map_pair.new_;
+          if (res.sm_.heartbeat_sm().smHeader.flags & RTPS::OPENDDS_FLAG_R) {
```

Comment:
We should only call `update_required_acknack_count` once when the count mapping assignment happens (after map_pair.new_ is assigned) for directed heartbeats. Where it's at now, the function will get called multiple times for the same heartbeat submessage, if it exists in multiple bundles. And it doesn't make sense to call it for undirected heartbeats, right (which it will also currently do).